File: backend/src/main/java/com/luisgosampaio/adventura/domain/group/Group.java
package com.luisgosampaio.adventura.domain.group;

import com.luisgosampaio.adventura.domain.user.User;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "group")
public class Group {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "name")
    private String name;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "created_by", nullable = false)
    private Long createdById;

    @OneToMany(mappedBy = "group", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<GroupMember> members = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PreUpdate
    protected void onUpdate () {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/types/group.ts
import type { User } from "./user";

export type GroupMemberRole = "ADMIN" | "MEMBER";

export interface GroupMember {
  id: number;
  group?: Group; // absent when serialized inside Group.members (@JsonBackReference)
  user: User;
  role: GroupMemberRole;
  enteredWhen: string;
}

export interface Group {
  id: number;
  name: string;
  description: string;
  createdBy: User;
  members: GroupMember[];
  createdAt: string;
  updatedAt: string;
}

export interface CreateGroupRequest {
  name: string;
  description: string;
}
